### PR TITLE
XEP-0191: Blocking Command implementation

### PR DESCRIPTION
**Pull request by ramabit**
Documentation: http://xmpp.org/extensions/xep-0191.html

### CODE REVIEW

File: smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+        if (blockingCommandManager == null) {
+            blockingCommandManager = new BlockingCommandManager(connection);
+            INSTANCES.put(connection, blockingCommandManager);
+        }
+
+        return blockingCommandManager;
+    }
+
+    private BlockingCommandManager(XMPPConnection connection) {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java]: I don't think we need the full Manager pattern for BlockingCommand. It's such a simple XEP and the fact that the constructor does nothing further indicates that this is not needed. Simply make all methods static and add another parameter `XMPPConnection` to them. Look for at ClientStateIndicationManager, IIRC it does the same.

Comment (ramabit) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java]: @Flowdalic done

File: smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+     * @throws NoResponseException
+     * @throws XMPPErrorException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     */
+    public static List<Jid> getBlockingList(XMPPConnection connection)
+            throws NoResponseException, XMPPErrorException, NotConnectedException, InterruptedException {
+        GetBlockListIQ getBlockListIQ = new GetBlockListIQ();
+        IQ responseIq = connection.createPacketCollectorAndSend(getBlockListIQ).nextResultOrThrow();
+        BlockListIQ blockListIQ = (BlockListIQ) responseIq;
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java]: FYI: You don't need this extra line. nextResultOrThrow() is generic and does the cast for you. But you don't need to change it if you don't want to. Just wanted to point this out.

Comment (ramabit) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java]: Ok thanks :)

File: smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/provider/BlockContactIQProvider.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+public class BlockContactIQProvider extends IQProvider<BlockContactIQ> {
+
+    @Override
+    public BlockContactIQ parse(XmlPullParser parser, int initialDepth) throws Exception {
+        Jid jid = null;
+
+        outerloop: while (true) {
+            int eventType = parser.next();
+
+            if (eventType == XmlPullParser.START_TAG) {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/provider/BlockContactIQProvider.java]: This should at least use switch-case. Ideally the parser name comparison below too, even if it has only one case at the moment.

**Comment by Flowdalic** [comment]
From a data model POV `BlockListIQ` and `GetBlockListIQ` are the exact same IQs. No need to have seem in different classes. Please merge them, into `BlocklistIQ`.

**Comment by Flowdalic** [comment]
Uhh, I just read a bit of XEP-191 and noticed that there is a blocklist push mechanism, similar to roster pushes, which you did not implement. Please implement that too, but this also means that you have to use the full manager idiom. Of course this also means that `getBlocklist()` - please rename the method, a "blockinglist" is something else - can work with  cached values.

**Comment by ramabit** [comment]
@Flowdalic How should I implement that push mechanism? Give some help

**Comment by Flowdalic** [comment]
> @Flowdalic How should I implement that push mechanism? Give me some help please

Register an IQ handler for the push. Then save the pushed blocking list in the manager. If the user calls `getBlocklist()` then first look if we already have one cached, only if not request the list from the server. Drop the cache if the connection goes down. Have a look at the privacy list manager, which does have the exact same caching mechanism, just for privacy lists.

**Comment by ramabit** [comment]
@Flowdalic ^ push IQ handler implemented :)
The continuous integration check throws an error in a file that I didn't change

File: smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    /**
+     * Returns the block list.
+     * 
+     * @return the blocking list
+     * @throws NoResponseException
+     * @throws XMPPErrorException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     */
+    public List<Jid> getBlockList()
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java]: The views handed out to the user of the blocklist should be unmodifiable. Please wrap the blocklist with `Collections.unmodifiableList()`.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+            throws NoResponseException, XMPPErrorException, NotConnectedException, InterruptedException {
+
+        if (blockListCached != null) {
+            return blockListCached;
+        }
+
+        BlockListIQ blockListIQ = new BlockListIQ(null);
+        blockListIQ.setType(Type.get);
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java]: IQs are per default of type `get`. This LOC can be removed.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+     * @throws InterruptedException
+     */
+    public List<Jid> getBlockList()
+            throws NoResponseException, XMPPErrorException, NotConnectedException, InterruptedException {
+
+        if (blockListCached != null) {
+            return blockListCached;
+        }
+
+        BlockListIQ blockListIQ = new BlockListIQ(null);
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java]: Please add a non-parameter constructor for `BlockListIQ` and use this constructor here.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java (hunk 4 above)
Comment (ramabit) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java]: Done

File: smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java (hunk 5 above)
Comment (ramabit) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java]: Done

File: smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java (hunk 6 above)
Comment (ramabit) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java]: Done

File: smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+     * @throws NoResponseException
+     * @throws XMPPErrorException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     */
+    public List<Jid> getBlockList()
+            throws NoResponseException, XMPPErrorException, NotConnectedException, InterruptedException {
+
+        if (blockListCached != null) {
+            return blockListCached;
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java]: This still returns a modifiable view to internal state of the manager.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        }
+
+        BlockListIQ blockListIQ = new BlockListIQ();
+        BlockListIQ blockListIQResult = connection().createPacketCollectorAndSend(blockListIQ).nextResultOrThrow();
+        blockListCached = blockListIQResult.getJids();
+        return (blockListCached == null) ? null : Collections.unmodifiableList(blockListCached);
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java]: Methods like these should never return `null` when they could return just an empty list. This removes the burden from the caller to handle the empty list case by checking for `null`.
But this is not only true for the high lever parts of the API, i.e. here, but also for the lower ones: `BlockListIQ.getJids()` should also never return `null`, but `Collections.emptyList()`.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java (hunk 7 above)
Comment (ramabit) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java]: Done

File: smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java (hunk 8 above)
Comment (ramabit) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/blocking/BlockingCommandManager.java]: Done

**Comment by Flowdalic** [comment]
Thanks, please squash and update the PR. I'll take over from here.

**Comment by ramabit** [comment]
@Flowdalic done :)

**Comment by Flowdalic** [comment]
Sorry I forget to assign an issue to this. Please update the commit so that it contains SMACK-731 somewhere within its commit message.

**Comment by ramabit** [comment]
@Flowdalic done

**Comment by ramabit** [comment]
@Flowdalic the automatic checks are not working ok since the last addition ("VersionEye")

**Comment by Flowdalic** [comment]
Did you loose the provider configuration for the blocking command providers in the `.providers` file when squashing the commits?

**Comment by Flowdalic** [comment]
Hmm it appears that your code does not allow for multiple items in <block/>. From XEP-0191: "The <block/> element MAY contain more than one <item/> child."

**Comment by ramabit** [comment]
@Flowdalic
1) The providers configuration is there. Check the tab "Files changed"
2) Ok, I will add support for multiple items

**Comment by ramabit** [comment]
@Flowdalic multiple items support added

**Comment by Flowdalic** [comment]
Ok, then please move the code from smack-experimental into smack-extensions, because -experimental is for experimental XEPs which XEP-0191 is not. This means that you must move the entry in `extenions.md` under the correct section too, and please make sure to add the entry in the right position, they are ordered by their XEP number. When done, please squash and update this PR.

**Comment by ramabit** [comment]
@Flowdalic done

**Comment by Flowdalic** [comment]
Needs squashing

**Comment by ramabit** [comment]
@Flowdalic squash done

**Comment by ramabit** [comment]
@Flowdalic the Travis CI is not working fine

**Comment by Flowdalic** [comment]
> @Flowdalic the Travis CI is not working fine

Yep, I know. I think I've already explained to you that this is because some unit tests are not properly synchronized. Which makes them fail sometimes, especially when executed within low processing power environments like Travis CI.

**Comment by ramabit** [comment]
@Flowdalic ok good to know, so I don't worry about that, hopefully everything seems to be ready for a merge ;)

**Comment by Flowdalic** [comment]
Closed with 9554c7be26344baf2412c78b6bdf5fc18f0c5e6a and 98bacb144d126bfbcbee41650a49e1f3b1077d90

**Comment by guusdk** [comment]
@ramabit Feel free to submit a patch for https://issues.igniterealtime.org/browse/OF-827 too :)

**Pull request closed without merging**